apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/profile/ProfileTaskExecutionContext.java

```diff
@@ -30,22 +37,139 @@
+            return false;
+        }
+
+        // check first operation name matches
+        if (!Objects.equals(task.getFistSpanOPName(), firstSpanOPName)) {
+            return false;
+        }
+
+        return tryToAttemptProfiling(tracingContext, traceSegmentId, firstSpanOPName, usingSlotCount);
+    }
+
+
+    /**
+     * profiling recheck
+     * @param tracingContext
+     * @param traceSegmentId
+     * @param firstSpanOPName
+     * @return
+     */
+    public boolean profilingRecheck(TracingContext tracingContext, ID traceSegmentId, String firstSpanOPName) {
+        boolean alreadyProfiling = tracingContext.isProfiling();
+
+        // not profiling and not available slot don't check anymore
+        int usingSlotCount = currentProfilingCount.get();
+        if (!alreadyProfiling && usingSlotCount >= Config.Profile.MAX_PARALLEL) {
+            return false;
+        }
+
+        // check first operation name matches
+        if (!Objects.equals(task.getFistSpanOPName(), firstSpanOPName)) {
+            if (alreadyProfiling) {
+                if (stopTracingProfile(tracingContext)) {
+                    // reduce total started profiling count when status is profiling
+                    totalStartedProfilingCount.addAndGet(-1);
+                }
+            }
+            return false;
+        } else if (alreadyProfiling) {
+            return true;
+        }
+
+        // not profiling, try to occupy slot
+        return tryToAttemptProfiling(tracingContext, traceSegmentId, firstSpanOPName, usingSlotCount);
+    }
+
+    /**
+     * find tracing context and clear on slot
+     *
+     * @param tracingContext
+     *
+     * @return current profiler is already start profiling
+     */
+    public boolean stopTracingProfile(TracingContext tracingContext) {
+        // find current tracingContext and clear it
+        boolean isProfilingStarted = false;
+        for (int slot = 0; slot < profilingSegmentSlots.length; slot++) {
+            ThreadProfiler currentProfiler = profilingSegmentSlots[slot];
+            if (currentProfiler != null && currentProfiler.matches(tracingContext)) {
+                profilingSegmentSlots[slot] = null;
+
+                // setting stop running
+                isProfilingStarted = currentProfiler.stopProfiling();
+                currentProfilingCount.addAndGet(-1);
+
+                profilingSegmentSlots = profilingSegmentSlots;
+                break;
+            }
+        }
+
+        return isProfilingStarted;
     }
 
     public ProfileTask getTask() {
         return task;
     }
 
-    public long getStartTime() {
-        return startTime;
+    public ThreadProfiler[] threadProfilerSlots() {
+        return profilingSegmentSlots;
     }
 
+    public boolean isStartProfileable(ThreadProfiler profiler) {
+        // check is out of max sampling count check
+        if (totalStartedProfilingCount.incrementAndGet() > task.getMaxSamplingCount()) {
```

Comment:
Why does this check happen at the start stage? I think when `totalStartedProfilingCount` reaches max count, this profile should be at the profiling stage. Do I miss anything?

Reply:
You are right, the check happens on the profiling stage. Maybe name as you feel confused, I change the name to `isProfileable`. It will invoke on `ThreadProfiler#startProfilingIfNeed`. When the tracing context starts time bigger than profile task `minDurationThresold`, It will have a check.